### ISSUE TITLE
Add support for `dyn Trait`

### ISSUE THREAD

**Issue by Nadrieril** [AeneasVerif/charon#123]

**Comment by Nadrieril** [comment]
I investigated a bit. The issue is that we need to handle [ExistentialPredicate](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_type_ir/predicate/enum.ExistentialPredicate.html) which isn't quite a `TraitRef`. If I understand, after https://github.com/AeneasVerif/charon/issues/127 we might be able to use a `TraitRef` there.

**Comment by sonmarcho** [comment]
What do you mean by "we might be able to use a `TraitRef` there?

**Comment by Nadrieril** [comment]
`dyn Trait` refers to a trait with an [ExistentialPredicate](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_type_ir/predicate/enum.ExistentialPredicate.html), which seems to be a mapping from `Self` to a list of clauses. E.g. `dyn Iterator<Item=u32>` would be represented as `lambda Self -> Self: Iterator, Self::Item = u8`. If we do #127, I _think_ this simplifies to just a list of trait clauses on `Self`, which we can represent as a list `TraitRef`s.

**Comment by sonmarcho** [comment]
I think a dyn trait really existentially quantifies a type, and we need to represent that: `lambda Self -> Self: Iterator, Self::Item = u8` is a predicate over some type `Self`, but must itself be quantified. If you have a function `fn foo(...) -> dyn Trait` in Lean we would generate something of the shape:
```lean
def foo ... : exists T, Trait T
```
or, written with a dependent tuple (we return a pair of a type `T` and the fact that it satisfies `Trait T`):
```lean
def foo ... : (T:Type, Trait T)
```

**Comment by Nadrieril** [comment]
Yes, but with #127 the existentially quantified predicate will always be of the form `lambda Self -> Self: Trait1, Self: Trait2`, which we can represent as a list `[Trait1, Trait2]`

**Comment by Nadrieril** [comment]
I was wrong about that: a `TraitRef` looks like `T: Into<u64>`, but what we need for `dyn` is just the `Into<u64>` part. In other words, it's a `TraitRef` where the `Self` argument is not supplied. In yet other words, it's an existentially quantified `TraitRef`.

**Comment by Nadrieril** [comment]
Partial support here: https://github.com/AeneasVerif/charon/pull/276. Needs more work to support fully.

**Comment by Nadrieril** [comment]
Technical note for myself: to create a `TraitRef` from a `TyKind::Dynamic`:
```rust
let ty::Dynamic(data, _, _) = *ty.kind() else { panic!(); }
let existential_trait_ref = tcx.instantiate_bound_regions_with_erased(data.principal().unwrap());
let trait_ref = existential_trait_ref.with_self_ty(tcx, ty);
```
(seen mentioned [here](https://rust-lang.zulipchat.com/#narrow/channel/144729-t-types/topic/Get.20list.20of.20dyn.20methods.20for.20a.20trait/near/492147123) in passing). We may want to handle regions better as well.
